src/Runtime/Core/Collections/IntPair.cs

```diff
@@ -0,0 +1,98 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+
+namespace Microsoft.ML.Probabilistic.Collections
+{
+    /// <summary>
+    /// Hand-rolled implementation of <see cref="ValueTuple{T1,T2}"/> for integers.
+    /// </summary>
+    /// <remarks>
+    /// <see cref="ValueTuple{T1,T2}.Equals(ValueTuple{T1,T2})"/> uses
+    /// <see cref="EqualityComparer{T}.Default"/> to implement he comparison. This actually involves
+    /// a lookup in a hash-table each time it is used due to the way static fields in generic
+    /// classes are implemented. It appears to be measurably slower when int pair is used as
```

Comment:
This line gives the impression that the `IntPair` is the slow one, though the opposite is supposed to be true.

Reply:
Will rewrite